test/test_common/logging.h

```diff
@@ -61,30 +61,103 @@ class LogRecordingSink : public Logger::SinkDelegate {
   std::vector<std::string> messages_;
 };
 
-// Validates that when stmt is executed, exactly one log message containing substr will be emitted.
-#define EXPECT_LOG_CONTAINS(loglevel, substr, stmt)                                                \
+typedef std::vector<std::pair<std::string, std::string>> ExpectedLogMessages;
+
+// Validates that when stmt is executed, log messages containing substr and loglevel will be
+// emitted. Failure message e.g.,
+//
+// Logs:
+// [2018-04-12 05:51:00.245][7290192][debug][upstream] grpc_mux_impl.cc:160] Received gRPC
+// [2018-04-12 05:51:00.246][7290192][warning][upstream] grpc_mux_impl.cc:63] Called bar
+// [2018-04-12 05:51:00.246][7290192][trace][upstream] grpc_mux_impl.cc:80] Sending foo
+//  Does NOT contain:
+//    'warning', 'Too many sendDiscoveryRequest calls for baz’
+//    'warning', 'Too man sendDiscoveryRequest calls for foo'
+#define EXPECT_LOG_CONTAINS_MANY(expected_messages, stmt)                                          \
```

Comment:
"MANY" seems ambiguous.  From reading the impl, I suggest calling this EXPECT_LOG_CONTAINS_ALL_OF.

And this is a nice abstraction as a test may not care about order or whether there are some extra logs beyond the expected ones.

Reply:
EXPECT_LOG_CONTAINS_ALL_OF is much better 👍